Assorted trivial cleanups 2/2019

mcostalba opened the pull request:
No functional change.

mcostalba commented:
This one https://github.com/official-stockfish/Stockfish/pull/2006/commits/0f6d4a1a8de13574d45038736e4145bc5defa450 is a bit of a speed up for me, but eventually an independent verification would be nice.

mstembera commented:
@mcostalba I measure over a 9% slowdown for 32-bit compiles w/ just the PopCnt16[] change.
make build ARCH=x86-32 CXX=i686-w64-mingw32-c++
```
Results for 40 tests for each version:

            Base      Test      Diff      
    Mean    1447560   1306987   140573    
    StDev   45907     46231     61824     

p-value: 0.011
speedup: -0.097
```

mstembera commented:
For the ForwardRanksBB[] change I do confirm a 1.5% speedup tested using the bmi2 compile.
```
Results for 40 tests for each version:

            Base      Test      Diff      
    Mean    1940386   1968693   -28307    
    StDev   71472     70351     97286     

p-value: 0.614
speedup: 0.015
```

31m059 commented:
You may also want to move Outpost so these Scores are in alphabetical order again.  Current master:
```c++
  // Assorted bonuses and penalties
  constexpr Score BishopPawns        = S(  3,  7);
  constexpr Score CorneredBishop     = S( 50, 50);
  constexpr Score FlankAttacks       = S(  8,  0);
  constexpr Score Hanging            = S( 69, 36);
  constexpr Score KingProtector      = S(  7,  8);
  constexpr Score KnightOnQueen      = S( 16, 12);
  constexpr Score LongDiagonalBishop = S( 45,  0);
  constexpr Score MinorBehindPawn    = S( 18,  3);
  constexpr Score PawnlessFlank      = S( 17, 95);
  constexpr Score RestrictedPiece    = S(  7,  7);
  constexpr Score RookOnPawn         = S( 10, 32);
  constexpr Score SliderOnQueen      = S( 59, 18);
  constexpr Score ThreatByKing       = S( 24, 89);
  constexpr Score ThreatByPawnPush   = S( 48, 39);
  constexpr Score ThreatByRank       = S( 13,  0);
  constexpr Score ThreatBySafePawn   = S(173, 94);
  constexpr Score TrappedRook        = S( 47,  4);
  constexpr Score WeakQueen          = S( 49, 15);
  constexpr Score WeakUnopposedPawn  = S( 12, 23);
  constexpr Score Outpost            = S(  9,  3);
```

mcostalba commented:
@mstembera thanks for testing. I have switched to a 32bit friendly version. Hopefully is a bit faster.

protonspring commented:
For my 64-bit builds, the 32bit friendly popcount is 10% slower than the original PopCnt16[] arrays.  Under what scenario is it faster to remove these?

protonspring commented:
In bitboard.h, line 248:  instead of "return x < y ? y - x : x - y;" we could do "std::abs(x - y);"
This is cleaner and seems to be just as fast as master.

mstembera commented:
@mcostalba With the 32 bit friendly version I get
```
Results for 40 tests for each version:

            Base      Test      Diff      
    Mean    1457178   1403334   53844     
    StDev   38598     45774     57180     

p-value: 0.173
speedup: -0.037
```
The ARCH=x86-64 compile is now 5% slower also.
```
Results for 40 tests for each version:

            Base      Test      Diff      
    Mean    1862458   1768653   93805     
    StDev   73207     70211     110567    

p-value: 0.198
speedup: -0.05
```

mstembera commented on src/bitboard.h:
```diff
@@ -192,48 +198,49 @@ inline Bitboard adjacent_files_bb(File f) {
 inline Bitboard forward_ranks_bb(Color c, Square s) {
-  return ForwardRanksBB[c][rank_of(s)];
+  return c == WHITE ? (AllSquares ^ Rank1BB) << 8 * rank_of(s)
+                    : (AllSquares ^ Rank8BB) >> 8 * (7 - rank_of(s));
```
What about removing AllSquares as well?
```C++
return c == WHITE ? ~Rank1BB << 8 * rank_of(s)
                  : ~Rank8BB >> 8 * (7 - rank_of(s));
```

mcostalba commented:
@protonspring @mstembera  Thanks! I have reverted PopCnt16. Indeed I think our tests are a bit unfair because in *real* 32bit CPU the available cache is much lower than your modern CPU, just compiled in 32 bit, and PopCnt16[] is a 64KB array. Anyhow it has little sense to spend time on this very specific and obscure side issue: I have just reverted it.

I have also added your additional suggestions, thanks!

protonspring commented:
I propose:
https://github.com/protonspring/Stockfish/tree/ps_promotions2

protonspring commented:
In ExtMove struct (movegen.h) add an operator:
bool operator==(ExtMove const& rhs) { return rhs.move == move; }

Then remove all of the .move's in movepick.cpp line 183+.

protonspring commented:
These few simplifications pass -3,1 with ease.

http://tests.stockfishchess.org/tests/view/5c6de6390ebc5925cffcbe72

31m059 commented:
Here's another one:

evaluate.cpp, lines 457-463:
```c++
    // Enemy knights checks
    b = pos.attacks_from<KNIGHT>(ksq) & attackedBy[Them][KNIGHT];

    if (b & safe)
        kingDanger += KnightSafeCheck;
    else
        unsafeChecks |= b;
```

`b` should probably be `KnightCheck` for consistency with RookCheck, QueenCheck, and BishopCheck, e.g.,
```c++
    // Enemy rooks checks
    Bitboard RookCheck =  b1
                        & safe
                        & attackedBy[Them][ROOK];

    if (RookCheck)
        kingDanger += RookSafeCheck;
    else
        unsafeChecks |= b1 & attackedBy[Them][ROOK];
```

 `b` then can be removed from line 405
```c++
    Bitboard weak, b, b1, b2, safe, unsafeChecks = 0;
```
as well.

Rocky640 commented:
in evaluate.cpp
```
 // Bonus for restricting their piece moves
    restricted =   attackedBy[Them][ALL_PIECES]
                & ~stronglyProtected
                &  attackedBy[Us][ALL_PIECES];
    score += RestrictedPiece * popcount(restricted);
```

use `b` instead of` restricted`,  as everywhere else.

mcostalba commented:
@protonspring thanks for the suggestion, but I'd would avoid to introduce heterogeneous compare operators, where operator==() compares on the move field and operator<() on the value field. This IMHO is a bit of confusion, better to leave the code explicit as is.

@31m059 @Rocky640 Thanks for the suggestions!

Vizvezdenec commented:
Well this slight mess with checks was result of my latest patch - because it touched 3 checks that can overlap (bishop, queen and rook) and didn't touch knight checks that can't overlap. 
For consistency having it named KnightCheck is probably a good thing.
Sure, we can have smth like b1 = rookcheck, b2 = queencheck, b = bishopcheck and then rewrite b but it will be hard to read I guess.

Vizvezdenec commented:
I have an idea to rename "initiative" evaluation function.
This eval term in SF is function of endgame score and is calculated in order to increase/decrease endgame evaluation depending on complexity of position.
Imho, it's not "initiative". I will call it something like "endgame winnability".
Current naming is especially bad for new developers since they try to improve it w/o understanding _that it's pure endgame bonus_ and try to add some midgame stuff there (I also fell in this trap quite a few times).

xoto10 commented:
I'm not sure it has to stay as endgame only though? As a matter of style I
would prefer it to get the full score as an input.

I view it as a bonus for the winning side to help them win more reliably.
This is currently endgame only but might add midgame elements one day.


Another, different way of looking at it, we evaluate;
the white position,
the black position,
the whole board (initiative)

Initiative is mostly looking at characteristics of the whole board, not
just black or white. I'm not sure if this is important, but it's one way of
characterizing what it does at the moment.

Vizvezdenec commented:
Sure but usually "initiative" term in chess is color-dependant.
"Initiative in a chess position belongs to the player who can make threats that cannot be ignored, thus putting the opponent in the position of having to spend turns responding to threats rather than creating new threats."
https://en.wikipedia.org/wiki/Initiative_(chess)
Right now "initiative" in SF is eval term that says how "winnable" position is for strong side - this basically has nothing to do with usual definition of it.
Even term itself is called "complexity" which is what it actually is more or less. But eval term is called "initiative" and this is somewhat of a miss-definition, IMHO.

31m059 commented:
@Vizvezdenec @xoto10 I fell in this trap a few times too.  I would suggest `complexity()` as the correct balance of

- more accurate naming, and more intuitive for people who are not deeply familiar with the code
- good continuity with previous SF versions (since the main term therein is already named "complexity") for the ease of frequent and longtime SF developers

protonspring commented:
In pawns::init(), we could remove this conditional?:

v += (Seed[r] + (phalanx ? (Seed[r + 1] - Seed[r]) / 2 : 0)) >> opposed ;

to

v += (Seed[r] + phalanx * (Seed[r + 1] - Seed[r]) / 2) >> opposed;

mcostalba commented:
Merged in 2bceba7f51621988
